web-api/src/storage/elastic/filter.rs

```diff
@@ -14,37 +14,96 @@
 
 use itertools::Itertools;
 use serde_json::{json, Value};
-use xayn_web_api_shared::serde::JsonObject;
+use xayn_web_api_shared::serde::{json_array, JsonObject};
 
 use crate::{
     personalization::filter::{CombineOp, CompareOp, Filter},
     storage::KnnSearchParams,
 };
 
+const BOOL: &str = "bool";
 const FILTER: &str = "filter";
 const MINIMUM_SHOULD_MATCH: &str = "minimum_should_match";
+const RANGE: &str = "range";
 const SHOULD: &str = "should";
 const TERM: &str = "term";
 const TERMS: &str = "terms";
 
-fn extend_value(filter: &mut JsonObject, occurence: &'static str, clause: Value) {
+fn merge_range(clause: &mut Vec<Value>) {
+    fn is_range(clause: &Value) -> Option<&str> {
+        clause
+            .as_object()
+            .and_then(|clause| clause.get(RANGE))
+            .and_then(|range| {
+                range
+                .as_object()
+                .unwrap(/* clause[RANGE] is object */)
+                .keys()
+                .next(/* clause[RANGE] contains exactly one key */)
+                .map(String::as_str)
+            })
+    }
+
+    let mut i = 0;
+    while i < clause.len() {
```

Comment:
but this will panic. in the inner while loop the `clause` gets shorter everytime we `swap_remove()` a range clause. so once we are again in the outer loop, `clause.len()` may have changed and `clause[i]` can panic.